### PR TITLE
General sync improvements

### PR DESCRIPTION
**Pull request by noffle**
# General Sync Improvements

This PR makes several changes. It would have been nice to keep them all separate, but many ended up being fix-as-you-go and aren't easily separable. I'll try to outline the major pieces to look for:

1. It introduces a "sync progress" suite of tests, using a new test helper function to do the heavy lifting of configuring several peers with different test data (map data & media), doing different sync combinations. It includes some tests that are marked as `test.skip()`, because they test progress features that are upcoming, but not yet implemented.
2. `mapeo-core#close()` now also halts indexing, so that the Node process can terminate properly.
3. **Many** bugs in the tests are fixed, especially around proper cleanup, isolated syncing, reliably syncing, and typos. This includes some very weird cases where peers would connect to each other across different tests on the same machine, as well as *between machines*.
4. A fix to `discovery-swarm` that prevents good sync connections from being terminated by new connections from the same machine of another type (UTPv4, UTPv6, TCPv4, TCPv6). Often machines will make several connections of some permutation of these. Now we only get one `peer` event per remote peer. Regardless of the # of connections, the first one is kept.
5. Some improvements to sync progress, such as not reporting any progress events until all incoming feeds' lengths are known.
6. Some cleanup to the `WifiPeer` object we pass around, to make it safer to work with. Before we were actually passing around a live patched version of `discovery-swarm`'s peer object!

tl;dr: Travis tests for sync now seem to be significantly more reliable, and we have a good suite of sync progress tests for the upcoming improvements to how progress gets reported.

### CODE REVIEW

**Comment by noffle** [comment]
Only took four months, but this PR is ready to go. :rocket:

**Comment by noffle** [comment]
Merging this is blocked by https://github.com/digidem/mapeo-desktop/issues/268

**Comment by okdistribute** [comment]
Does this PR also depend upon hypercore8?

**Comment by noffle** [comment]
In the sense of "can this be backported"?

I don't *think so*, not explicitly. However, it might be quite
challenging to backport, since `sync.js` has changed since 8.x.x and may
not accept the patch so readily.
